src/main/drivers/serial_uart.c

```diff
@@ -297,6 +297,67 @@ static void uartWrite(serialPort_t *instance, uint8_t ch)
     }
 }
 
+static void uartBeginWrite(serialPort_t *instance)
+{
+    uartPort_t *uartPort = (uartPort_t *)instance;
+
+    // Check if the TX line is being pulled low by an unpowered peripheral
+    if (uartPort->checkUsartTxOutput && !uartPort->checkUsartTxOutput(uartPort)) {
+        // TX line is being pulled low, so don't transmit
+        return;
+    }
+}
+
+static void uartWriteBuf(serialPort_t *instance, const void *data, int count)
+{
+    uartPort_t *uartPort = (uartPort_t *)instance;
+    uartDevice_t *uart = container_of(uartPort, uartDevice_t, port);
+    const uint8_t *bytePtr = (const uint8_t*)data;
+
+    // Check if the TX line is being pulled low by an unpowered peripheral
+    if (uartPort->checkUsartTxOutput && (uart->txPinState != TX_PIN_ACTIVE)) {
+        // TX line is being pulled low, so don't transmit
+        return;
+    }
+
+    for (int i = 0; i < count; i++) {
+        uartPort->port.txBuffer[uartPort->port.txBufferHead] = *bytePtr++;
+
+        if (uartPort->port.txBufferHead + 1 >= uartPort->port.txBufferSize) {
+            uartPort->port.txBufferHead = 0;
+        } else {
+            uartPort->port.txBufferHead++;
+        }
+    }
```

Comment:
```suggestion
while (count > 0) {
    // Calculate the available space to the end of the buffer
    int spaceToEnd = uartPort->port.txBufferSize - uartPort->port.txBufferHead;
    // Determine the amount to copy in this iteration: the minimum of 'count' or 'spaceToEnd'
    int chunkSize = (count < spaceToEnd) ? count : spaceToEnd;

    // Copy the chunk
    memcpy(&uartPort->port.txBuffer[uartPort->port.txBufferHead], bytePtr, chunkSize);

    // Update pointers and counters
    bytePtr += chunkSize; // Advance source pointer
    uartPort->port.txBufferHead = (uartPort->port.txBufferHead + chunkSize) % uartPort->port.txBufferSize; // Wrap or advance head
    count -= chunkSize; // Decrease remaining count
}

```

chatGPT is a bit chatty, but code is correct and faster

Reply:
Yeah, that's good. First use of AI in Betaflight to my knowledge! I've adjusted the comments a bit.